src/common/KokkosKernels_Sorting.hpp

```diff
@@ -0,0 +1,631 @@
+    {
+      //Count the number of elements in each bucket
+      Ordinal count[16] = {0};
+      Ordinal offset[17];
+      if(!inAux)
+      {
+        for(Ordinal i = 0; i < n; i++)
+        {
+          count[(values[i] & mask) >> maskPos]++;
+        }
+      }
+      else
+      {
+        for(Ordinal i = 0; i < n; i++)
+        {
+          count[(valuesAux[i] & mask) >> maskPos]++;
+        }
+      }
+      offset[0] = 0;
+      //get offset as the prefix sum for count
+      for(Ordinal i = 0; i < 16; i++)
+      {
+        offset[i + 1] = offset[i] + count[i];
+      }
+      //now for each element in [lo, hi), move it to its offset in the other buffer
+      //this branch should be ok because whichBuf is the same on all threads
+      if(!inAux)
+      {
+        //copy from *Over to *Aux
+        for(Ordinal i = 0; i < n; i++)
+        {
+          Ordinal bucket = (values[i] & mask) >> maskPos;
+          valuesAux[offset[bucket + 1] - count[bucket]] = values[i];
+          count[bucket]--;
+        }
+      }
+      else
+      {
+        //copy from *Aux to *Over
+        for(Ordinal i = 0; i < n; i++)
+        {
+          Ordinal bucket = (valuesAux[i] & mask) >> maskPos;
+          values[offset[bucket + 1] - count[bucket]] = valuesAux[i];
+          count[bucket]--;
+        }
+      }
+      inAux = !inAux;
+      mask = mask << 4;
+      maskPos += 4;
+    }
+  });
+  //Move values back into main array if they are currently in aux.
+  //This is the case if an odd number of rounds were done.
+  if(((sortBits + 3) / 4) % 2 == 1)
+  {
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(thread, n),
+    [=](Ordinal i)
+    {
+      values[i] = valuesAux[i];
+    });
+  }
+  //remove bias to restore original values
+  if(minVal != 0)
+  {
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(thread, n),
+    [=](Ordinal i)
+    {
+      values[i] += minVal;
+    });
+  }
+}
+
+//Radix sort for integers, on a single thread within a team.
+//While sorting, also permute "perm" array along with the values.
+//Pros: few diverging branches, so good for sorting on a single GPU thread/warp.
+//Con: requires auxiliary storage, this version only works for integers (although float/double is possible)
+template<typename Ordinal, typename ValueType, typename PermType, typename Thread>
+KOKKOS_INLINE_FUNCTION void
+radixSort2(ValueType* values, ValueType* valuesAux, PermType* perm, PermType* permAux, Ordinal n, const Thread thread)
+{
+  static_assert(std::is_integral<ValueType>::value, "radixSort can only be run on integers.");
+  if(n <= 1)
+    return;
+  ValueType minVal = Kokkos::ArithTraits<ValueType>::max();
+  ValueType maxVal = 0;
+  for(Ordinal i = 0; i < n; i++)
+  {
+    if(minVal > values[i])
+      minVal = values[i];
+    if(maxVal < values[i])
+      maxVal = values[i];
+  }
+  //apply a bias so that key range always starts at 0
+  //also invert key values here for a descending sort
```

Comment:
Will this work on Volta ? I thought you would need a barrier here, as you use the minVal rightaway. May be not because minVal is threadlocal ?

Reply:
Yes, everything inside radixSort/radixSort2 is thread-local, so I don't think warp divergence can cause any issues.